File: nodejs/src/services/PersonaService.ts
import { PersonaDto, CreatePersonaDto, UpdatePersonaDto, PersonaSearchDto } from '../dto/PersonaDto';
import { PersonaAtributoDto } from '../dto/PersonaAtributoDto';
import { IPersonaService } from './IPersonaService';
import { IPersonaRepository } from '../repositories/IPersonaRepository';
import { Persona } from '../entities/Persona';

export class PersonaService implements IPersonaService {
  constructor(private personaRepository: IPersonaRepository) {}

  async findAll(): Promise<PersonaDto[]> {
    const personas = await this.personaRepository.findAll();
    return personas.map(this.mapToDto);
  }

  async findById(id: number): Promise<PersonaDto | null> {
    const persona = await this.personaRepository.findById(id);
    return persona ? this.mapToDto(persona) : null;
  }

  async create(createPersonaDto: CreatePersonaDto): Promise<PersonaDto> {
    // Validar que la identificación sea única
    if (await this.personaRepository.existsByIdentificacion(createPersonaDto.identificacion)) {
      throw new Error('Ya existe una persona con esta identificación');
    }

    const persona = new Persona();
    persona.nombreCompleto = createPersonaDto.nombreCompleto;
    persona.identificacion = createPersonaDto.identificacion;
    persona.edad = createPersonaDto.edad;
    persona.genero = createPersonaDto.genero;
    persona.estado = createPersonaDto.estado || 'Activo';
    persona.atributosAdicionales = JSON.stringify(createPersonaDto.atributosAdicionales || {});

    const createdPersona = await this.personaRepository.create(persona);
    return this.mapToDto(createdPersona);
  }

  async update(id: number, updatePersonaDto: UpdatePersonaDto): Promise<PersonaDto> {
    const persona = await this.personaRepository.findById(id);
    if (!persona) {
      throw new Error(`No se encontró la persona con ID: ${id}`);
    }

    // Validar que la identificación sea única (excluyendo el registro actual)
    if (await this.personaRepository.existsByIdentificacion(updatePersonaDto.identificacion, id)) {
      throw new Error('Ya existe otra persona con esta identificación');
    }

    persona.nombreCompleto = updatePersonaDto.nombreCompleto;
    persona.identificacion = updatePersonaDto.identificacion;
    persona.edad = updatePersonaDto.edad;
    persona.genero = updatePersonaDto.genero;
    persona.estado = updatePersonaDto.estado;
    persona.atributosAdicionales = JSON.stringify(updatePersonaDto.atributosAdicionales || {});

    const updatedPersona = await this.personaRepository.update(persona);
    return this.mapToDto(updatedPersona);
  }

  async delete(id: number): Promise<void> {
    const persona = await this.personaRepository.findById(id);
    if (!persona) {
      throw new Error(`No se encontró la persona con ID: ${id}`);
    }

    await this.personaRepository.delete(id);
  }

  async search(searchDto: PersonaSearchDto): Promise<PersonaDto[]> {
    const personas = await this.personaRepository.search(
      searchDto.nombre,
      searchDto.estado,
      searchDto.edadMinima,
      searchDto.edadMaxima
    );
    return personas.map(this.mapToDto);
  }

  async validateIdentificacion(identificacion: string, excludeId?: number): Promise<boolean> {
    return !(await this.personaRepository.existsByIdentificacion(identificacion, excludeId));
  }

  private mapToDto(persona: Persona): PersonaDto {
    let atributosAdicionales: Record<string, any> = {};
    try {
      if (persona.atributosAdicionales) {
        atributosAdicionales = JSON.parse(persona.atributosAdicionales);
      }
    } catch {
      // Si hay error al parsear, usar objeto vacío
    }

    return {
      id: persona.id,
      nombreCompleto: persona.nombreCompleto,
      identificacion: persona.identificacion,
      edad: persona.edad,
      genero: persona.genero,
      estado: persona.estado,
      fechaCreacion: persona.fechaCreacion,
      fechaModificacion: persona.fechaModificacion,
      atributosAdicionales,
      atributosDetallados: persona.personaAtributos?.map(pa => ({
        id: pa.id,
        atributoTipoId: pa.atributoTipoId,
        nombreAtributo: pa.atributoTipo?.nombre || '',
        tipoDato: pa.atributoTipo?.tipoDato || '',
        valor: pa.valor,
        esObligatorio: pa.atributoTipo?.esObligatorio || false
      })) || []
    };
  }
}

File: dotnet/PersonasABM.Application/Services/PersonaService.cs
using System.Text.Json;
using Microsoft.Extensions.Logging;
using PersonasABM.Application.DTOs;
using PersonasABM.Application.Interfaces;
using PersonasABM.Domain.Entities;

namespace PersonasABM.Application.Services;

public class PersonaService : IPersonaService
{
    private readonly IPersonaRepository _personaRepository;
    private readonly ILogger<PersonaService> _logger;

    public PersonaService(IPersonaRepository personaRepository, ILogger<PersonaService> logger)
    {
        _personaRepository = personaRepository;
        _logger = logger;
    }

    public async Task<IEnumerable<PersonaDto>> GetAllAsync()
    {
        _logger.LogInformation("Obteniendo todas las personas");
        var personas = await _personaRepository.GetAllAsync();
        return personas.Select(MapToDto);
    }

    public async Task<PersonaDto?> GetByIdAsync(int id)
    {
        _logger.LogInformation("Obteniendo persona con ID: {Id}", id);
        var persona = await _personaRepository.GetByIdAsync(id);
        return persona != null ? MapToDto(persona) : null;
    }

    public async Task<PersonaDto> CreateAsync(CreatePersonaDto createPersonaDto)
    {
        _logger.LogInformation("Creando nueva persona: {Identificacion}", createPersonaDto.Identificacion);
        
        // Validar que la identificación sea única
        if (await _personaRepository.ExistsByIdentificacionAsync(createPersonaDto.Identificacion))
        {
            throw new InvalidOperationException("Ya existe una persona con esta identificación");
        }

        var persona = new Persona
        {
            NombreCompleto = createPersonaDto.NombreCompleto,
            Identificacion = createPersonaDto.Identificacion,
            Edad = createPersonaDto.Edad,
            Genero = createPersonaDto.Genero,
            Estado = createPersonaDto.Estado,
            AtributosAdicionales = JsonSerializer.Serialize(createPersonaDto.AtributosAdicionales)
        };

        var createdPersona = await _personaRepository.AddAsync(persona);
        _logger.LogInformation("Persona creada exitosamente con ID: {Id}", createdPersona.Id);
        
        return MapToDto(createdPersona);
    }

    public async Task<PersonaDto> UpdateAsync(int id, UpdatePersonaDto updatePersonaDto)
    {
        _logger.LogInformation("Actualizando persona con ID: {Id}", id);
        
        var persona = await _personaRepository.GetByIdAsync(id);
        if (persona == null)
        {
            throw new KeyNotFoundException($"No se encontró la persona con ID: {id}");
        }

        // Validar que la identificación sea única (excluyendo el registro actual)
        if (await _personaRepository.ExistsByIdentificacionAsync(updatePersonaDto.Identificacion, id))
        {
            throw new InvalidOperationException("Ya existe otra persona con esta identificación");
        }

        persona.NombreCompleto = updatePersonaDto.NombreCompleto;
        persona.Identificacion = updatePersonaDto.Identificacion;
        persona.Edad = updatePersonaDto.Edad;
        persona.Genero = updatePersonaDto.Genero;
        persona.Estado = updatePersonaDto.Estado;
        persona.AtributosAdicionales = JsonSerializer.Serialize(updatePersonaDto.AtributosAdicionales);
        persona.FechaModificacion = DateTime.UtcNow;

        var updatedPersona = await _personaRepository.UpdateAsync(persona);
        _logger.LogInformation("Persona actualizada exitosamente con ID: {Id}", updatedPersona.Id);
        
        return MapToDto(updatedPersona);
    }

    public async Task DeleteAsync(int id)
    {
        _logger.LogInformation("Eliminando persona con ID: {Id}", id);
        
        var persona = await _personaRepository.GetByIdAsync(id);
        if (persona == null)
        {
            throw new KeyNotFoundException($"No se encontró la persona con ID: {id}");
        }

        await _personaRepository.DeleteAsync(id);
        _logger.LogInformation("Persona eliminada exitosamente con ID: {Id}", id);
    }

    public async Task<IEnumerable<PersonaDto>> SearchAsync(PersonaSearchDto searchDto)
    {
        _logger.LogInformation("Buscando personas con filtros: {SearchDto}", JsonSerializer.Serialize(searchDto));
        
        var personas = await _personaRepository.SearchAsync(
            searchDto.Nombre, 
            searchDto.Estado, 
            searchDto.EdadMinima, 
            searchDto.EdadMaxima);
        
        return personas.Select(MapToDto);
    }

    public async Task<bool> ValidateIdentificacionAsync(string identificacion, int? excludeId = null)
    {
        return !await _personaRepository.ExistsByIdentificacionAsync(identificacion, excludeId);
    }

    private static PersonaDto MapToDto(Persona persona)
    {
        var atributosAdicionales = new Dictionary<string, object>();
        try
        {
            if (!string.IsNullOrEmpty(persona.AtributosAdicionales))
            {
                atributosAdicionales = JsonSerializer.Deserialize<Dictionary<string, object>>(persona.AtributosAdicionales) ?? new();
            }
        }
        catch
        {
            // Si hay error al deserializar, usar diccionario vacío
        }

        return new PersonaDto
        {
            Id = persona.Id,
            NombreCompleto = persona.NombreCompleto,
            Identificacion = persona.Identificacion,
            Edad = persona.Edad,
            Genero = persona.Genero,
            Estado = persona.Estado,
            FechaCreacion = persona.FechaCreacion,
            FechaModificacion = persona.FechaModificacion,
            AtributosAdicionales = atributosAdicionales,
            AtributosDetallados = persona.PersonaAtributos.Select(pa => new PersonaAtributoDto
            {
                Id = pa.Id,
                AtributoTipoId = pa.AtributoTipoId,
                NombreAtributo = pa.AtributoTipo.Nombre,
                TipoDato = pa.AtributoTipo.TipoDato,
                Valor = pa.Valor,
                EsObligatorio = pa.AtributoTipo.EsObligatorio
            }).ToList()
        };
    }
}
